src/dscanner/analysis/objectconst.d

```diff
@@ -5,100 +5,89 @@
-import dparse.lexer;
 import dscanner.analysis.base;
 import dscanner.analysis.helpers;
-import dsymbol.scope_ : Scope;
+import std.stdio;
 
-/**
- * Checks that opEquals, opCmp, toHash, 'opCast', and toString are either const,
- * immutable, or inout.
- */
-final class ObjectConstCheck : BaseAnalyzer
+extern(C++) class ObjectConstCheck(AST) : BaseAnalyzerDmd!AST
 {
-	alias visit = BaseAnalyzer.visit;
-
 	mixin AnalyzerInfo!"object_const_check";
+	alias visit = BaseAnalyzerDmd!AST.visit;
 
-	///
-	this(string fileName, const(Scope)* sc, bool skipTests = false)
+	extern(D) this(string fileName)
 	{
-		super(fileName, sc, skipTests);
+		super(fileName);
 	}
 
-	mixin visitTemplate!ClassDeclaration;
-	mixin visitTemplate!InterfaceDeclaration;
-	mixin visitTemplate!UnionDeclaration;
-	mixin visitTemplate!StructDeclaration;
-
-	override void visit(const AttributeDeclaration d)
+	void visitAggregate(AST.AggregateDeclaration ad)
 	{
-		if (d.attribute.attribute == tok!"const" && inAggregate)
-		{
-			constColon = true;
-		}
-		d.accept(this);
-	}
+		import dmd.astenums : MODFlags, STC;
 
-	override void visit(const Declaration d)
-	{
-		import std.algorithm : any;
-		bool setConstBlock;
-		if (inAggregate && d.attributes && d.attributes.any!(a => a.attribute == tok!"const"))
-		{
-			setConstBlock = true;
-			constBlock = true;
-		}
-
-		bool containsDisable(A)(const A[] attribs)
-		{
-			import std.algorithm.searching : canFind;
-			return attribs.canFind!(a => a.atAttribute !is null &&
-				a.atAttribute.identifier.text == "disable");
-		}
+		if (!ad.members)
+			return;
 
-		if (const FunctionDeclaration fd = d.functionDeclaration)
+		foreach(member; *ad.members)
 		{
-			const isDeclationDisabled = containsDisable(d.attributes) ||
-				containsDisable(fd.memberFunctionAttributes);
-
-			if (inAggregate && !constColon && !constBlock && !isDeclationDisabled
-					&& isInteresting(fd.name.text) && !hasConst(fd.memberFunctionAttributes))
+			if (auto fd = member.isFuncDeclaration())
+			{
+				if (fd.type && !fd.type.mod == MODFlags.const_ && isInteresting(fd.ident.toString()) && 
+					!(fd.storage_class & STC.disable))
+						addErrorMessage(cast(ulong) fd.loc.linnum, cast(ulong) fd.loc.charnum, KEY,
+							"Methods 'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' are non-const.");
+				
+				member.accept(this);
+			}
+			else if (auto scd = member.isStorageClassDeclaration())
 			{
-				addErrorMessage(d.functionDeclaration.name.line,
-						d.functionDeclaration.name.column, "dscanner.suspicious.object_const",
-						"Methods 'opCmp', 'toHash', 'opEquals', 'opCast', and/or 'toString' are non-const.");
+				foreach (smember; *scd.decl)
+				{
+					if (auto fd2 = smember.isFuncDeclaration())
+					{
+						if (fd2.type && !fd2.type.mod == MODFlags.const_ && isInteresting(fd2.ident.toString()) && 
+							!(fd2.storage_class & STC.disable) && !(scd.stc & STC.const_ ||
+							scd.stc & STC.immutable_ || scd.stc & STC.wild))
```

Comment:
Why is this condition different from the one above? You might want to consider putting this condition in a separate function.

Reply:
This condition is different because the storage class is saved in the StorageClassDeclaration class, and not in the storage_class field of a FuncDeclaration. Why do you think it is needed to add this condition in a separate function? And are we talking only about this condition, or the other condition as well? (from isFuncDeclaration() block)

Reply:
Except for the last part, the condition is the same: type not null, type not const, function is interesting, function not disabled. You could create a function:

```d
bool isConstFunction(FuncDeclaration fd, StorageClassDeclaration scd = null)
{
     /*...*/
}